File: frontend/src/types/transaction.ts
export interface Category {
  id: number;
  nome: string;
  tipo: string;
  cor?: string;
  icone?: string;
}

export interface BudgetAlertInfo {
  tipo_alerta: "aviso" | "estouro";
  budget_id: string;
  budget_nome: string;
  categoria_id: number;
  categoria_nome: string;
  valor_limite: number;
  valor_gasto: number;
  valor_disponivel: number;
  percentual_gasto: number;
  percentual_notificacao: number;
  periodo_info: {
    ano: number;
    mes: number;
    quinzena?: number;
    semana?: number;
    data_inicio: string;
    data_fim: string;
    periodicidade: string;
  };
  alerta_enviado: boolean;
}

export interface Transaction {
  id: string;
  usuario_id: string;
  mensagem_original: string;
  valor: number;
  descricao: string;
  tipo: "despesa" | "receita";
  canal?: "audioMessage" | "conversation" | "imageMessage" | "webApp";
  categoria_id?: number;
  data_transacao: string; // ISO date string
  data_registro: string; // ISO datetime string
  categoria?: Category;
  budget_alert?: BudgetAlertInfo;
}

export interface TransactionCreate {
  mensagem_original: string;
  valor: number;
  descricao: string;
  tipo: "despesa" | "receita";
  canal?: "audioMessage" | "conversation" | "imageMessage";
  categoria_id?: number;
  data_transacao?: string; // ISO date string
}

export interface TransactionUpdate {
  descricao?: string;
  valor?: number;
  categoria_id?: number;
  data_transacao?: string; // ISO date string
}

export interface TransactionFilters {
  tipo?: "despesa" | "receita";
  categoria_id?: number;
  data_inicio?: string; // ISO date string
  data_fim?: string; // ISO date string
  valor_min?: number;
  valor_max?: number;
  search?: string; // Search in description
}

export interface TransactionStats {
  total_receitas: number;
  total_despesas: number;
  saldo: number;
  receitas: number; // count
  despesas: number; // count
}

export interface CategorySummary {
  categoria: string;
  tipo: string;
  total: number;
  quantidade: number;
}

export interface PaginatedTransactions {
  items: Transaction[];
  total: number;
  page: number;
  size: number;
  pages: number;
}

export interface TransactionListParams {
  page?: number;
  size?: number;
  tipo?: "despesa" | "receita";
  categoria_id?: number;
  data_inicio?: string;
  data_fim?: string;
  search?: string;
}

File: backend/app/schemas/transaction.py
from datetime import date, datetime
from decimal import Decimal
from typing import List, Literal, Optional
from uuid import UUID

from pydantic import BaseModel


class TransactionBase(BaseModel):
    mensagem_original: str
    valor: Decimal
    descricao: str
    tipo: Literal["despesa", "receita"]
    canal: Optional[
        Literal["audioMessage", "conversation", "imageMessage", "webApp"]
    ] = None
    categoria_id: Optional[int] = None
    data_transacao: Optional[date] = None


class TransactionCreate(TransactionBase):
    usuario_id: UUID


class TransactionUpdate(BaseModel):
    descricao: Optional[str] = None
    valor: Optional[Decimal] = None
    categoria_id: Optional[int] = None
    data_transacao: Optional[date] = None


class TransactionInDB(TransactionBase):
    id: UUID
    usuario_id: UUID
    data_registro: datetime

    class Config:
        from_attributes = True


class Transaction(TransactionInDB):
    pass


class TransactionWithCategory(Transaction):
    categoria: Optional["Category"] = None
    budget_alert: Optional[dict] = None  # Informações de alerta de orçamento, se houver

    class Config:
        from_attributes = True


# Additional schemas for API endpoints
class TransactionFilters(BaseModel):
    """Filtros para listagem de transações"""

    tipo: Optional[Literal["despesa", "receita"]] = None
    categoria_id: Optional[int] = None
    data_inicio: Optional[date] = None
    data_fim: Optional[date] = None


class TransactionStats(BaseModel):
    """Estatísticas de transações"""

    total_receitas: Decimal
    total_despesas: Decimal
    saldo: Decimal
    receitas: int
    despesas: int


class CategorySummary(BaseModel):
    """Resumo por categoria"""

    categoria_id: int
    categoria_nome: str
    tipo: str
    total_valor: Decimal
    total_transacoes: int


class PaginatedTransactions(BaseModel):
    """Transações paginadas"""

    items: List[TransactionWithCategory]
    total: int
    page: int
    size: int
    pages: int


# Import here to avoid circular imports
from .category import Category  # noqa: E402

TransactionWithCategory.model_rebuild()
